flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksNativeFullSnapshotStrategy.java

```diff
@@ -0,0 +1,410 @@
+import java.util.Map;
+import java.util.UUID;
+
+import static org.apache.flink.contrib.streaming.state.snapshot.RocksSnapshotUtil.getUploadedStateSize;
+
+/**
+ * Snapshot strategy for {@link RocksDBKeyedStateBackend} based on RocksDB's native checkpoints and
+ * creates full snapshots. the difference between savepoint is that sst files will be uploaded
+ * rather than states.
+ *
+ * @param <K> type of the backend keys.
+ */
+public class RocksNativeFullSnapshotStrategy<K>
+        extends RocksDBSnapshotStrategyBase<
+                K, RocksNativeFullSnapshotStrategy.NativeFullRocksDBSnapshotResources> {
+
+    private static final Logger LOG =
+            LoggerFactory.getLogger(RocksNativeFullSnapshotStrategy.class);
+
+    private static final String DESCRIPTION = "Asynchronous full RocksDB snapshot";
+
+    /** The state handle ids of all sst files materialized in snapshots for previous checkpoints. */
+    @Nonnull private final UUID backendUID;
+
+    /** The help class used to upload state files. */
+    private final RocksDBStateUploader stateUploader;
+
+    public RocksNativeFullSnapshotStrategy(
+            @Nonnull RocksDB db,
+            @Nonnull ResourceGuard rocksDBResourceGuard,
+            @Nonnull TypeSerializer<K> keySerializer,
+            @Nonnull LinkedHashMap<String, RocksDbKvStateInfo> kvStateInformation,
+            @Nonnull KeyGroupRange keyGroupRange,
+            @Nonnegative int keyGroupPrefixBytes,
+            @Nonnull LocalRecoveryConfig localRecoveryConfig,
+            @Nonnull File instanceBasePath,
+            @Nonnull UUID backendUID,
+            @Nonnull RocksDBStateUploader rocksDBStateUploader) {
+        super(
+                DESCRIPTION,
+                db,
+                rocksDBResourceGuard,
+                keySerializer,
+                kvStateInformation,
+                keyGroupRange,
+                keyGroupPrefixBytes,
+                localRecoveryConfig,
+                instanceBasePath,
+                backendUID.toString().replaceAll("[\\-]", ""));
+        this.backendUID = backendUID;
+        this.stateUploader = rocksDBStateUploader;
+    }
+
+    @Override
+    public NativeFullRocksDBSnapshotResources syncPrepareResources(long checkpointId)
+            throws Exception {
+
+        final SnapshotDirectory snapshotDirectory = prepareLocalSnapshotDirectory(checkpointId);
+        LOG.trace("Local RocksDB checkpoint goes to backup path {}.", snapshotDirectory);
+
+        final List<StateMetaInfoSnapshot> stateMetaInfoSnapshots =
+                new ArrayList<>(kvStateInformation.size());
+
+        for (Map.Entry<String, RocksDbKvStateInfo> stateMetaInfoEntry :
+                kvStateInformation.entrySet()) {
+            stateMetaInfoSnapshots.add(stateMetaInfoEntry.getValue().metaInfo.snapshot());
+        }
+
+        takeDBNativeCheckpoint(snapshotDirectory);
+
+        return new NativeFullRocksDBSnapshotResources(snapshotDirectory, stateMetaInfoSnapshots);
```

Comment:
These lines are almost the same as `RocksIncrementalSnapshotStrategy`, and I think we could move such classes (e.g. `NativeFullRocksDBSnapshotResources`) and methods (e.g. `#materializeMetaData`) to `RocksDBSnapshotStrategyBase` to avoid code duplication.